diff -Naur cqrlog-1.9.0.orig/Makefile cqrlog-1.9.0/Makefile
--- cqrlog-1.9.0.orig/Makefile	2015-03-04 09:17:21.000000000 -0600
+++ cqrlog-1.9.0/Makefile	2015-03-06 21:17:31.002138573 -0600
@@ -6,14 +6,12 @@
 
 cqrlog: src/cqrlog.lpi
 	$(CC) --ws=gtk2 src/cqrlog.lpi
-	$(ST) src/cqrlog
-	gzip tools/cqrlog.1 -c > tools/cqrlog.1.gz
 
 clean:
 	rm -f -v src/*.o src/*.ppu src/*.bak src/lnet/lib/*.ppu src/lnet/lib/*.o src/lnet/lib/*.bak src/cqrlog src/cqrlog.compiled debian/cqrlog.* src/ipc/*.o src/ipc/*.ppu src/cqrlog.or
 	rm -rf debian/cqrlog
 	rm -f -v src/mysql/*.ppu src/mysq/*.bak src/mysql/*.o
-	rm -f -v tools/cqrlog.1.gz
+	rm -f -v tools/cqrlog.1
 	
 install:
 	install -d -v         $(bindir)
@@ -44,9 +42,8 @@
 	install    -v -m 0644 images/*   $(datadir)/images/
 	install    -v -m 0644 tools/cqrlog.desktop $(sharedir)/applications/cqrlog.desktop
 	install    -v -m 0644 images/cqrlog.png $(sharedir)/pixmaps/cqrlog/cqrlog.png
-	install    -v -m 0644 images/cqrlog.png $(sharedir)/icons/cqrlog.png  
 	install    -v -m 0644 src/changelog.html $(datadir)/changelog.html
-	install    -v -m 0644 tools/cqrlog.1.gz $(sharedir)/man/man1/cqrlog.1.gz
+	install    -v -m 0644 tools/cqrlog.1 $(sharedir)/man/man1/cqrlog.1
 deb:
 	dpkg-buildpackage -rfakeroot -i -I
 deb_src:
