diff -up ./src/dData.pas.orig ./src/dData.pas
--- ./src/dData.pas.orig	2018-06-17 05:42:42.000000000 -0700
+++ ./src/dData.pas	2018-10-26 15:07:05.053119031 -0700
@@ -4221,7 +4221,11 @@ begin
     Paths.Add('/lib/');
     Paths.Add('/usr/lib/mysql/');
 
-    Result := MyFindFile('libmariadbclient.so*', Paths);
+    Result := MyFindFile('libmariadb.so*', Paths);
+    if (Result='') then
+    begin
+      Result := MyFindFile('libmariadbclient.so*', Paths)
+    end;
     if (Result='') then
     begin
       Result := MyFindFile('libmysqlclient.so*', Paths)
